### PR TITLE
Typed JavaScriptIndexCreationTask API

### LINKED ISSUE #208: Better DX for defining AbstractJavaScriptIndexCreationTask
This is more of a question/suggestion.

Looking at the tests for `AbstractJavaScriptIndexCreationTask` I'm wondering if we'll eventually be able to just express the index in raw JS since we can do `Function.toString()` to grab the raw definition to store. Could even provide some TypeScript static typing to help.

Example:

http://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAkgdgEwgDwCIQGYEs5eFgezgB4AVAPigF4oAKBAgYwFcBbCOYALilIEpqlAgCMAVhEbAAUFMYAbAIYBnJVACCwpcABOCyQCkFANwUBlRtqxhg8JMgDC2iAvxFSygNZQA3lKj+oMEsTYGgAfVYFMFUaOAgAdyhTCGBiLUs4AHNyWj4ZAMDtAlDJCAQoSLAyHMYCOTkJVzgedJxMgBooJGxcJp5bFHQevEISCgFfAoLgAAssJQA6CKjFhQQEWgADStoAcgASb1r6xtGAX13Ow+6cEaIF4AJTHTbcs75NvgBuPwCz-ICmRSFRWuR8vymUCcwGY2jgUFm8yWlSUPwK-3+UhwoW0GD00AAqkoINpwQVmMS4Qp2C0Xlk0QEsAhaRlMj9MfJlKoiSSlGEAEIgHlU9hQFChRCqDTpPTAQwmcyWawDBxOFyjdxKLyTAK1ODpZiSAjaME6yFKZiQE15CFTRGLSrEYU5XbCpSXOjMARUSi0bxQCkkuDUiA8ZgLQMi6DvG2Q-x6pR1CALOQETK0e0LFF5dFSDlELRQHB2ahQOKJN0CoWU4PsXJAA

### PR DESCRIPTION
**Pull request by yj7o5**
- Moved old AbstractJavaScriptIndexCreationTask into AbstractRawJavaScriptIndexCreationTask
- Added Typed JS API under AbstractJavaScriptIndexCreationTask
- Added typed Map interface that takes in Map definition as input
- Added typed Reduce interface that takes the groupBy and aggregate definition from the user vs.  setting `this.reduce` so that its more natural.

TODO: 
- [x] a bit more cleanup
- [x] few more test cases (such as for `createSpatialField` etc)
- [x] make sure prod build minification doesn't break anything, don't think it would unless we're obfuscating the build

closes #208

### CODE REVIEW

File: src/Documents/Indexes/AbstractJavaScriptIndexCreationTask.ts [hunk 1]
```diff
@@ -11,14 +11,6 @@ export class AbstractJavaScriptIndexCreationTask extends AbstractIndexCreationTa
         this._definition.priority = "Normal";
     }
 
-    public get maps() {
```
Comment (ml054) [inline src/Documents/Indexes/AbstractJavaScriptIndexCreationTask.ts:14]: why do we remove that from here?

File: src/Documents/Indexes/AbstractJavaScriptIndexCreationTask.ts [hunk 2]
```diff
@@ -27,16 +19,8 @@ export class AbstractJavaScriptIndexCreationTask extends AbstractIndexCreationTa
         this._definition.fields = value;
     }
 
-    public get reduce() {
```
Comment (ml054) [inline src/Documents/Indexes/AbstractJavaScriptIndexCreationTask.ts:30]: same here.

**Review by ml054 — APPROVED** [review]

**Comment by yj7o5** [comment]
@ml054 any idea on when this will merged into 5.0 and when the 5.0 client will be made available. Thanks.

**Comment by ml054** [comment]
Hi, 

The schedule goes like this:
- I need to finish sync changes from 5.0 -> node.js client 
- then we review API as 5.0 is great occasion for breaking changes
- will try to merge and test your PR - I need to expose all helper functions we have in queries. 

ETA. 4-6 weeks.

**Comment by ml054** [comment]
fixed as a part of: https://github.com/ravendb/ravendb-nodejs-client/pull/241

**Pull request closed without merging**